geopandas/io/file.py

```diff
@@ -148,27 +148,73 @@ def convert_type(column, in_type):
 
-    geom_type = _common_geom_type(df)
+    # Since https://github.com/Toblerity/Fiona/issues/446 resolution,
+    # Fiona allows a list of geometry types
+    geom_types = _geometry_types(df)
 
-    schema = {'geometry': geom_type, 'properties': properties}
+    schema = {'geometry': geom_types, 'properties': properties}
 
     return schema
 
 
-def _common_geom_type(df):
-    # Need to check geom_types before we write to file...
-    # Some (most?) providers expect a single geometry type:
-    # Point, LineString, or Polygon
-    geom_types = df.geometry.geom_type.unique()
+def _geometry_types(df):
+    geom_types = find_geometry_types(df)
 
-    from os.path import commonprefix
-    # use reversed geom types and commonprefix to find the common suffix,
-    # then reverse the result to get back to a geom type
-    geom_type = commonprefix([g[::-1] for g in geom_types if g])[::-1]
-    if not geom_type:
+    if len(geom_types) == 0:
+        # Default geometry type supported by Fiona
+        # (Since https://github.com/Toblerity/Fiona/issues/446 resolution)
         return 'Unknown'
 
+    if len(geom_types) == 1:
+        geom_types = geom_types[0]
+
+    return geom_types
+
+
+def find_geometry_types(df):
+    if _FIONA18:
+        # Starting from Fiona 1.8, schema submitted to fiona to write a gdf
+        # can have mixed geometries:
+        # - 3D and 2D shapes can coexist in inferred schema
+        # - Shape and MultiShape types can (and must) coexist in inferred schema
+        geom_types = _3D_geom_types(df) + _2D_geom_types(df)
+    else:
+        # Before Fiona 1.8, schema submitted to write a gdf should have
+        # one single geometry type whenever possible:
+        # - 3D and 2D shapes cannot coexist in inferred schema
+        # - Shape and MultiShape can not coexist in inferred schema
+        geom_types = _merge_geom_types_as_much_as_possible(df)
+
+    return geom_types
+
+
+def _2D_geom_types(df):
+    geom_types_2D = df[~df.geometry.has_z].geometry.geom_type.unique()
+    return [type for type in geom_types_2D if type is not None]
+
+
+def _3D_geom_types(df):
+    geom_types_3D = df[df.geometry.has_z].geometry.geom_type.unique()
+    return ["3D " + type for type in geom_types_3D if type is not None]
+
+
+def _merge_geom_types_as_much_as_possible(df):
+    """
+    for backward compatibility with Fiona<1.8 only
+    """
+    unique_geom_types = df.geometry.geom_type.unique()
+    unique_geom_types = [type for type in unique_geom_types if type is not None]
+    unique_geom_types = merge_shape_and_multi_shape_types(unique_geom_types)
     if df.geometry.has_z.any():
-        geom_type = "3D " + geom_type
+        # declare all geometries as 3D geometries
+        unique_geom_types = ["3D " + type for type in unique_geom_types]
+    # by default, all geometries are 2D geometries
+
+    return unique_geom_types
+
 
-    return geom_type
+def merge_shape_and_multi_shape_types(unique_geom_types):
```

Comment:
can you make this a private function as well?

Reply:
oops, my bad...
sure